api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java

```diff
@@ -217,17 +218,48 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
 				toAdd.add(newObs);
 			}
 		}
-		
-		for (Obs o : toRemove) {
+
+		removeObsWithGroupMembers(encounter, toRemove);
+		addObsWithGroupMembers(encounter, toAdd);
+		return encounter;
+	}
+
+	/**
+	 * This method will remove list of obs recursively from encounter
+	 *
+	 * @param encounter the encounter from which the obs will be removed
+	 * @param obsToRemove list of obs need to be removed recursively
+	 */
+	private void removeObsWithGroupMembers(Encounter encounter, List<Obs> obsToRemove) {
```

Comment:
The method name seems to suggest removing obs that have (with) group members.
Wouldn't removeObsAndGroupMembers be less confusing?